### PR TITLE
fix: Performance improvements and throttling features

### PR DESCRIPTION
**Pull request by untitaker**
Improve performance in dump_events.py, fix some crashes encountered with
production data and make sure we dive under the rate limits of our DSN.

### CODE REVIEW

File: upload_events.py [hunk 1]
```diff
@@ -61,31 +68,54 @@ def upload_events(file_name: Path, dsn: str, project_id: int, project_slug: str,
...
+                envelope.add_event(event)
+
+                q.put(envelope)
+
+                event_count += 1
+
+                if event_sleep and time.time() - last_sleep > 10 and event_count > 100:
+                    time.sleep(event_count * event_sleep / 1000.0)
+                    last_sleep = time.time()
+                    event_count = 0
```
Comment (RaduW) [inline upload_events.py:115]: Why do we do this ? (i.e. why do we need to sleep ? )

File: upload_events.py [hunk 2]
```diff
@@ -25,9 +31,10 @@
 @click.option("--project-id", type=int, help="the project id to be dumped, alternative to dsn, or (project-slug,org-slug)")
 @click.option("--project-slug", type=str, help="project slug, must also use org-slug, alternatively use project-id or dsn")
 @click.option("--org-slug", type=str, help="organization slug, must also use project-slug, alternatively use project-id or dsn")
-@click.option("--expire-days", default=0, type=int, help="sets timestamp in the past to influence retention")
+@click.option("--network-threads", default=64, help="How many threads to use for sending")
```
Comment (RaduW) [inline upload_events.py:34]: In my tests, 2 * num_cpu seemed to get the fastest processing time.

**Review by RaduW — APPROVED** [review]
I like the way we do batching.

In my tests dumping the events was very fast (compared with sending them), and didn't see any point in complicating the code with doing it on multiple threads. Probably the events you are using are a lot bigger.

File: upload_events.py (hunk 1 above)
Comment (untitaker) [reply upload_events.py:115]: throttling for rate limits. we're asked to send no more than 30 events per second

File: upload_events.py (hunk 2 above)
Comment (untitaker) [reply upload_events.py:34]: yeah I still have to tweak this number in prod. 64 doesn't harm really, and we need to throttle upload anyway.

**Pull request merged**